### PR TITLE
InputUniDriver - Unidriver migration/input

### PR DESCRIPTION
**Pull request by mosheKerbel**
Adding InputUniDriver and testing it with the existing input.spec.js

TODO:
1. Uncomment the tooltip related tests once the tooltip uniDriver is merged.
2. Complete the clear method tests.

### CODE REVIEW

File: src/Input/Input.unidriver.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      (await driver.getRootElementClasses()).indexOf(styles.disabled) !== -1,
+    isHoveredStyle: async () =>
+      (await driver.getRootElementClasses()).indexOf(styles.hasHover) !== -1,
+    isFocusedStyle: async () =>
+      (await driver.getRootElementClasses()).indexOf(styles.hasFocus) !== -1,
+    getRequired: async () => await reactBaseInput.required(),
+    enterText: async value => await input.enterValue(value),
+    getAutocomplete: async () => await input.attr('autocomplete'),
+    getDefaultValue: async () => await reactBaseInput.defaultValue(),
+    getUnit: async () => await reactBase.textContent(),
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: ### BUG
You forgot to query for the unit node first.
`element.querySelector(`.${styles.unit}`)`

"Luckily", the test is a bad one )-:
Try fixing it by adding a `suffix`:
```js
it('should the unit text if passed', async () => {
        const unit = '$';

        const { driver } = render(<Input suffix="Hello" unit={unit} />);
        expect(await driver.getUnit()).toEqual(unit);
      });
```

File: src/Input/Input.unidriver.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import { baseUniDriverFactory } from 'wix-ui-test-utils/base-driver';
+import styles from './Input.scss';
+import { ReactBase } from '../../test/utils/unidriver';
+
+export const testkit = (base,body,document) => {
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: This needs a refactor.
It should be `({base, document})`
So opened [this issue](https://github.com/wix-a/wsr-unidriver-migration/issues/11), so you don't have to do it now.

File: src/Input/Input.unidriver.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      (await driver.getRootElementClasses()).indexOf(styles.hasFocus) !== -1,
+    getRequired: async () => await reactBaseInput.required(),
+    enterText: async value => await input.enterValue(value),
+    getAutocomplete: async () => await input.attr('autocomplete'),
+    getDefaultValue: async () => await reactBaseInput.defaultValue(),
+    getUnit: async () => await reactBase.textContent(),
+    getTabIndex: async () => await reactBaseInput.tabIndex(),
+    getReadOnly: async () => await reactBaseInput.readOnly(),
+    getTextOverflow: async () =>
+      (await input.getNative()).style['text-overflow'],
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Don't use `getNative()`, Add to `ReactBase`

File: src/Input/Input.unidriver.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    hasExclamation: async () => await base.$(`.${styles.exclamation}`).exists(),
+    hasClass: async className =>
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Use unidriver's `hasClass`

File: src/Input/Input.unidriver.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      (await base.attr('class')).indexOf(className) !== -1,
+    hasError: async () => await driver.hasClass(styles.hasError),
+    hasLoader: async () =>
+      (await base.$$(`.${styles.loaderContainer}`).count()) > 0,
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Why not a single `$`, and use `exists()`?

File: src/Input/Input.unidriver.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    focus: async () => await reactBaseInput.focus(),
+    blur: async () => await reactBaseInput.blur(),
+    keyUp: async () => await reactBaseInput.keyup(),
+    keyDown: async () => await reactBaseInput.keydown(),
+    paste: async () => await reactBaseInput.paste(),
+    trigger: async (value, event) => {
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Whatever...

File: src/Input/Input.unidriver.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        return await driver.enterText(value);
+      }
+    },
+    isFocus: async () => await reactBaseInput.isFocus(),
+    hasHelp: async () => await base.$(`.${styles.help}`).exists(),
+    clickUnit: async () => await base.$(`.${styles.unit}`).click(),
+    hasMagnifyingGlass: async () =>
+      await base.$(`.${styles.magnifyingGlass}`).exists(),
+    clickMagnifyingGlass: async () =>
+      await base.$(`.${styles.magnifyingGlass}`).click(),
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Optionally extract: `base.$(`.${styles.magnifyingGlass}`)`

File: src/Input/Input.unidriver.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    clickMenuArrow: async () => await base.$(`.${styles.menuArrow}`).click(),
+    hasMenuArrow: async () => await base.$(`.${styles.menuArrow}`).exists(),
```
Comment (hepiyellow) [inline src/Input/Input.unidriver.js]: Optionally extract `base.$(`.${styles.narrow}`)`

File: test/utils/unidriver/ReactBase.js [hunk 9]
```diff
@@ -16,8 +16,20 @@ export function ReactBase(base) {
...
+    },
+    blur: async () => Simulate.blur(await htmlElement()),
+    keyup: async () => Simulate.keyUp(await htmlElement()),
+    keydown: async (key) => Simulate.keyDown(await htmlElement(), key),
+    paste: async () => Simulate.paste(await htmlElement()),
+    required: async () => (await htmlElement()).required,
+    defaultValue: async () => (await htmlElement()).defaultValue,
+    textContent: async () => (await htmlElement()).textContent,
+    selectionStart: async () => (await htmlElement()).selectionStart,
   };
```
Comment (hepiyellow) [inline test/utils/unidriver/ReactBase.js]: Missing methods:
- [ ] startComposing
- [ ] endComposing
- [ ] hasRightBorderRadius
- [ ] hasLeftBorderRadius
- [ ] isOfSize
- [ ] getCustomAffix
- [ ] getDataHook
- [ ] getTooltipDataHook
- [ ] getTooltipElement
- [ ] clearText
- [ ] mouseOut
- [ ] mouseOver

File: src/Input/Input.js [hunk 10]
```diff
@@ -351,6 +351,8 @@ class Input extends Component {
         ...event.target,
         value: '',
       };
+      //prevents the syntactic event from being nullified by React
+      event && event.persist && event.persist();
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/Input/Input.js]: Yes, this test was failing prior to the change: https://github.com/wix/wix-style-react/blob/e513d78ebebe9acf0f6ca36200484c349afc2ca0/src/Input/Input.spec.js#L589

File: test/utils/unidriver/ReactBase.js (hunk 9 above)
Comment (hepiyellow) [reply test/utils/unidriver/ReactBase.js]: @mosheKerbel 
So remind me what did we say about all these missing methods?

Cause now it seems to me that they are needed.
They are public so we don't want to break them.
Its true that for now we don't encourage consumers to use the new async drivers, but they ARE going to be available.

Comment (hepiyellow) [reply test/utils/unidriver/ReactBase.js]: I'll open an issue for these

Comment (mosheKerbel) [reply test/utils/unidriver/ReactBase.js]: We said we should create an issue for that
Did that yesterday
https://github.com/wix/wix-style-react/issues/3035

**Pull request merged**